### PR TITLE
Iox #86 introduce cxx function ref

### LINKED ISSUE #86: Introduce cxx::function_ref
## Brief feature description
Introduce a light-weight alternative `cxx::function_ref` as `std::function` can use heap under certain conditions which does not meet the requirements for high-safety environments

## Detailed information
`cx::function_ref` - a non-owning reference to a callable

Proposed features:
* Not use heap
* No exceptions
* Stateful lambda support
* C++11 support

Loosely based on this proposal: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0792r5.html

### PR DESCRIPTION
**Pull request by mossmaurice**
Resolves #86

### CODE REVIEW

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 1]
```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
```
Comment (elBoberido) [inline iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:1]: please move the code to an inl file

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+class function_ref<ReturnType(ArgTypes...)>
+{
+    using SignatureType = ReturnType(ArgTypes...);
+    using SelfType = function_ref<SignatureType>;
+    template <typename T>
+    using EnableIfNotFunctionRef = typename std::enable_if<!std::is_same<std::decay<T>, function_ref>::value>::type;
+
+  public:
+    /// @brief Creates an empty function_ref
+    function_ref() noexcept
```
Comment (elBoberido) [inline iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:63]: not really a big fan of this. optional<function_ref> would be better in my point of view, but it seems the decision was to do it this way

on the other hand, you would result in the same, invalid object after a move operation

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    /// @brief Checks whether a valid target is contained
+    explicit operator bool() const noexcept
+    {
+        return m_target != nullptr;
+    }
+
+    void swap(function_ref& rhs) noexcept
+    {
+        std::swap(m_target, rhs.m_target);
```
Comment (elBoberido) [inline iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:134]: is std::swap part of the ASILD subset of the STL we can use?

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  public:
+    /// @brief Creates an empty function_ref
+    function_ref() noexcept
+        : m_target(nullptr)
+        , m_functionPointer(nullptr)
+    {
+    }
+
+    /// @brief Creates an empty function_ref
+    function_ref(nullptr_t) noexcept
```
Comment (elBoberido) [inline iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:70]: wouldn't it be better to have it `std::nullptr_t`

Comment (elfenpiff) [inline iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:70]: What is the use case of this constructor?
```function_ref bla{nullptr}``` would construct the identical object like ```function_ref bla``` ?
And this constructor can only be used in combination with a nullptr ... why?

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp (hunk 2 above)
Comment (elfenpiff) [inline iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:63]: Please move the implementation into a separate ```function_ref.inl``` file

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    EXPECT_FALSE(sut);
+}
+
+TEST_F(function_refTest, CreateWithNullptr)
+{
+    function_ref<void()> sut(nullptr);
+    EXPECT_FALSE(sut);
+}
+
+TEST_F(function_refDeathTest, CreateEmptyLeadsToTermination)
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:92]: Better name would probably be CallEmptyLeadsToTermination

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    /// @brief Move assignment operator
+    function_ref& operator=(function_ref&& rhs) noexcept
+    {
+        if (this != &rhs)
+        {
+            m_target = rhs.m_target;
+            m_functionPointer = rhs.m_functionPointer;
+            // Make sure no UB can happen by marking the lvalue as invalid
```
Comment (elfenpiff) [inline iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:108]: I know UB is short for undefined behavior but I would write it out here

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+
+TEST_F(function_refTest, CreateEmptyAndAssign)
+{
+    auto lambda = [] {};
+    function_ref<void()> sut;
+    sut = lambda;
+    EXPECT_TRUE(sut);
+}
+
+TEST_F(function_refTest, CreateAndCopy)
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:106]: maybe CreateAndCopyConstruct

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+{
+    auto lambda = []() -> int { return 42; };
+    function_ref<int()> sut1(lambda);
+    function_ref<int()> sut2{sut1};
+    EXPECT_TRUE(sut2);
+    EXPECT_THAT(sut2(), Eq(42));
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:112]: please use a constexpr instead of magic numbers

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        }
+        return *this;
+    };
+
+    /// @brief Calls the provided callable
+    ReturnType operator()(ArgTypes... args) const noexcept
+    {
+        if (!m_target)
+        {
+            // Callable was called without user having assigned one beforehand
```
Comment (elfenpiff) [inline iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:120]: Maybe we should print out a fatal error message like
```
fprintf(stderr, "empty funtion_ref invoked, can't live with that shame, committing seppuku");
```

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+    auto lambda = []() -> int { return 42; };
+    function_ref<int()> sut2;
+    {
+        function_ref<int()> sut1(lambda);
+        EXPECT_THAT(sut1(), Eq(42));
+        EXPECT_FALSE(sut2);
+        sut2 = sut1;
+    }
+    EXPECT_TRUE(sut2);
+    EXPECT_THAT(sut2(), Eq(42));
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:126]: again, no magic numbers
also, this is a bit paranoid, but use another number. just in the case that for some weird reasons the copy assigns the address of the lambda to the lambda of the previous test which memory might not have been overwritten and therefore returns 42

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+
+struct ComplexType
+{
+    char a;
+    int b;
+    float c;
+
+    bool operator==(const ComplexType& rhs) const
+    {
+        if (a == rhs.a && b == rhs.b && c == rhs.c)
```
Comment (elfenpiff) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:43]: Why not write simply:
```
return ( a == rhs.a && b == rhs.b && c == rhs.c);
```
This would save you 7 lines and lines these days are very expensive ;)

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+    {
+        return 4273;
+    }
+
+    uint8_t m_iter{0};
+};
+
+using function_refDeathTest = function_refTest;
+
+TEST_F(function_refTest, CreateEmpty)
```
Comment (elfenpiff) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:80]: I would let the name of the tests be more detailled. Write what you are testing and what you are expecting. Like CreateEmptyIsFalse

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 7 above)
Comment (elfenpiff) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:106]: I would add previous to this test a test where you assign a lambda and call it since this is not tested

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+TEST_F(function_refTest, CreateAndCopy)
+{
+    auto lambda = []() -> int { return 42; };
+    function_ref<int()> sut1(lambda);
+    function_ref<int()> sut2{sut1};
+    EXPECT_TRUE(sut2);
+    EXPECT_THAT(sut2(), Eq(42));
+}
+
+TEST_F(function_refTest, CreateAndCopyAssign)
```
Comment (elfenpiff) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:115]: Maybe you should also add a tests where you copy a function_ref into an invalid/empty function or where you copy an invalid function_ref into a valid one.

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+        function_ref<int()> sut1(lambda);
+        EXPECT_THAT(sut1(), Eq(42));
+        EXPECT_FALSE(sut2);
+        sut2 = sut1;
+    }
+    EXPECT_TRUE(sut2);
+    EXPECT_THAT(sut2(), Eq(42));
+}
+
+TEST_F(function_refTest, CreateAndMove)
```
Comment (elfenpiff) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:129]: Here you should also add tests where you move a valid function ref into an invalid one and an invalid function_ref into a valid one

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+    function_ref<int()> sut1{lambda1};
+    {
+        function_ref<int()> sut2{lambda2};
+        sut1 = std::move(sut2);
+    }
+    EXPECT_TRUE(sut1);
+    EXPECT_THAT(sut1(), Eq(73));
+}
+
+TEST_F(function_refTest, CreateAndSwap)
```
Comment (elfenpiff) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:152]: We should also test swapping valid-invalid function_refs

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+{
+    Functor foo;
+    function_ref<int()> sut(foo);
+    EXPECT_THAT(sut(), Eq(73));
+}
+
+TEST_F(function_refTest, CreateWithStdBind)
+{
+    function_ref<int()> sut(std::bind(&function_refTest::foobar, this));
+    EXPECT_THAT(sut(), Eq(4273));
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:210]: no magic numbers please

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp (hunk 4 above)
Comment (elfenpiff) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:70]: And please write ```std::nullptr_t```

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp (hunk 3 above)
Comment (elfenpiff) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:134]: I dont think so but it should be feasible to implement this on our own

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp (hunk 6 above)
Comment (elBoberido) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:108]: maybe mossmaurice means unexpected behaviour or unspecified? Could you please clarify ;)

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 11 above)
Comment (elBoberido) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:43]: that's boring ... on the other side, Walter E. Brown knows much more interesting ways to return true ;)

Comment (mossmaurice) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:43]: If it is test code I don't care about LoC, but prefer to make it more readable :-) For everyone interested in the brilliant talk, here's the link: https://www.youtube.com/watch?v=OAmWHmwlMwI

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp (hunk 4 above)
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:70]: Done

Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:70]: Support different flavours is generally a good idea IMHO. `function_ref bla{nullptr}` is more readable to some folks.

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp (hunk 2 above)
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:63]: Done

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp (hunk 4 above)
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp]: Thinking about it again, it's a quite odd API. I'll remove it.

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 8 above)
Comment (mossmaurice) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:112]: Magic numbers a magic 🦄 For small tests I prefer to omit the extra variable.

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 10 above)
Comment (mossmaurice) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:126]: Made all numbers unique.

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 12 above)
Comment (mossmaurice) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:80]: You're right, rephrased test case names

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 7 above)
Comment (mossmaurice) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:106]: Added

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 13 above)
Comment (mossmaurice) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:115]: Added

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 14 above)
Comment (mossmaurice) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:129]: Added

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 15 above)
Comment (mossmaurice) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:152]: Added

File: iceoryx_utils/include/iceoryx_utils/internal/cxx/function_ref.inl [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+    return *this;
+};
+
+template <class ReturnType, class... ArgTypes>
+inline ReturnType function_ref<ReturnType(ArgTypes...)>::operator()(ArgTypes... args) const noexcept
+{
+    if (!m_target)
+    {
+        // Callable was called without user having assigned one beforehand
+        std::cerr << "Empty function_ref invoked" << std::endl;
```
Comment (budrus) [inline iceoryx_utils/include/iceoryx_utils/internal/cxx/function_ref.inl:65]: I think we have to come up with a strategy whether is is cool to directly use std::cerr or if we should use our logger. As this is not the first affected module, this PR is fine so far but I created a new issue to address this question #130

Comment (elBoberido) [reply iceoryx_utils/include/iceoryx_utils/internal/cxx/function_ref.inl:65]: the idea was to use cout, clog and cerr in the utils and then let the logger capture that and map it to a LogInfo, LogWarn and LogError ... unfortunatelly there was not yet time to do this

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 8 above)
Comment (elBoberido) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp]: I disagree, but not enough to prevent this merge

it should become muscle memory and that only happens if you do it all the time ;)

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+
+TEST_F(function_refTest, CreateValidByAssignIsTrue)
+{
+    auto lambda = [] {};
+    function_ref<void()> sut;
+    sut = lambda;
+    EXPECT_TRUE(sut);
+}
+
+TEST_F(function_refTest, CreateValidByAssignResultEqual)
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:104]: shouldn't this also be CallValid...

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+{
+    auto lambda = []() -> int { return 7253; };
+    function_ref<int()> sut;
+    sut = lambda;
+    EXPECT_THAT(sut(), Eq(7253));
+}
+
+TEST_F(function_refTest, CreateValidByCopyConstructResultEqual)
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:112]: again, CallValid...

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+{
+    auto lambda = []() -> int { return 3527; };
+    function_ref<int()> sut1{lambda};
+    function_ref<int()> sut2(sut1);
+    EXPECT_TRUE(sut2);
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:117]: please make this an ASSERT_TRUE, else the whole test will terminate in the next call

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 21]
```diff
@@ -0,0 +1,276 @@
...
+    }
+    EXPECT_FALSE(sut2);
+}
+
+TEST_F(function_refTest, CreateValidByMoveResultEqual)
+{
+    auto lambda = []() -> int { return 123; };
+    function_ref<int()> sut1{lambda};
+    function_ref<int()> sut2{std::move(sut1)};
+    EXPECT_TRUE(sut2);
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:152]: please make an ASSERT_THAT, else the access to sut2 further down might terminate the whole test

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 22]
```diff
@@ -0,0 +1,276 @@
...
+TEST_F(function_refTest, CreateValidByMoveAssignResultEqual)
+{
+    auto lambda1 = []() -> int { return 118; };
+    auto lambda2 = []() -> int { return 999; };
+    function_ref<int()> sut1{lambda1};
+    {
+        function_ref<int()> sut2{lambda2};
+        sut1 = std::move(sut2);
+    }
+    EXPECT_TRUE(sut1);
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:173]: Again, ASSERT_THAT

Comment (mossmaurice) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:173]: Will be done in #131

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 21 above)
Comment (mossmaurice) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:152]: Good point, will be done in #131

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 20 above)
Comment (mossmaurice) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:117]: Will be done in follow-up #131

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 19 above)
Comment (mossmaurice) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:112]: I'll rename it just for you <3

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 11 above)
Comment (elfenpiff) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:47]: I disagree. How does it make your code more readable when you just add more lines?
The first line ```( a == rhs.a && b == rhs.b && c == rhs.c)``` have to be understood either way and with your approach you have 6 more lines to understood

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+///         // Usage as function parameter
+///         void fuu(cxx::function_ref<void()> callback)
+///         {
+///             callback();
+///         }
+///         // Call the lambda
+///         fuu([]{ doSomething(); });
+///
+///         // Usage with l-values
+///         // Pitfall: Ensure that lifetime of callable suits the point in time of calling callback()
```
Comment (elfenpiff) [inline iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:44]: This pitfall has to be stated in the doxygen docu in the method

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+
+template <class ReturnType, class... ArgTypes>
+class function_ref<ReturnType(ArgTypes...)>
+{
+    using SignatureType = ReturnType(ArgTypes...);
+    template <typename T>
+    using EnableIfNotFunctionRef = typename std::enable_if<!std::is_same<std::decay<T>, function_ref>::value>::type;
+
+  public:
+    /// @brief Creates an empty function_ref
```
Comment (elfenpiff) [inline iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:62]: You should maybe state the default constructor creates a function_ref in an invalid state and operator bool() returns false

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+    function_ref() noexcept;
+
+    /// @brief D'tor
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:65]: Touché

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+    ~function_ref() noexcept = default;
+
+    /// @brief Copy c'tor
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:68]: Your Karma point +1 ;)

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+    function_ref(const function_ref&) noexcept = default;
+
+    /// @brief Copy assignment operator
+    function_ref& operator=(const function_ref&) noexcept = default;
+
+    /// @brief Create a function_ref
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:74]: Added

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
+    template <typename CallableType, typename = EnableIfNotFunctionRef<CallableType>>
+    function_ref(CallableType&& callable) noexcept;
+
+    /// @brief Moves a function_ref
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:78]: Done

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 29]
```diff
@@ -0,0 +1,105 @@
...
+    function_ref(function_ref&& rhs) noexcept;
+
+    /// @brief Move assignment operator
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:81]: Done

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 30]
```diff
@@ -0,0 +1,105 @@
...
+    function_ref& operator=(function_ref&& rhs) noexcept;
+
+    /// @brief Calls the provided callable
+    ReturnType operator()(ArgTypes... args) const noexcept;
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:85]: Done

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 31]
```diff
@@ -0,0 +1,105 @@
...
+
+    /// @brief Checks whether a valid target is contained
+    explicit operator bool() const noexcept;
+
+    void swap(function_ref& rhs) noexcept;
+
+  private:
+    /// @brief Raw pointer of the callable
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:93]: Valid point, renamed it.

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 32]
```diff
@@ -0,0 +1,105 @@
...
+    /// @brief Moves a function_ref
+    function_ref(function_ref&& rhs) noexcept;
+
+    /// @brief Move assignment operator
+    function_ref& operator=(function_ref&& rhs) noexcept;
+
+    /// @brief Calls the provided callable
+    ReturnType operator()(ArgTypes... args) const noexcept;
+
+    /// @brief Checks whether a valid target is contained
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:87]: Done

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 33]
```diff
@@ -0,0 +1,105 @@
...
+    explicit operator bool() const noexcept;
+
+    void swap(function_ref& rhs) noexcept;
+
+  private:
+    /// @brief Raw pointer of the callable
+    void* m_target{nullptr};
+
+    /// @brief Function pointer to the callable
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:96]: Removed

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp [hunk 34]
```diff
@@ -0,0 +1,105 @@
...
+    /// @brief Move assignment operator
+    function_ref& operator=(function_ref&& rhs) noexcept;
+
+    /// @brief Calls the provided callable
+    ReturnType operator()(ArgTypes... args) const noexcept;
+
+    /// @brief Checks whether a valid target is contained
+    explicit operator bool() const noexcept;
+
+    void swap(function_ref& rhs) noexcept;
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:90]: Added documentation

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp (hunk 23 above)
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:44]: Done

File: iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp (hunk 24 above)
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp:62]: Added
